engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCompoundShape.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2017 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.physics.bullet.shapes;
+
+
+import com.bulletphysics.collision.shapes.CompoundShape;
+import com.bulletphysics.collision.shapes.CompoundShapeChild;
+import com.bulletphysics.linearmath.Transform;
+import org.terasology.math.Rotation;
+import org.terasology.math.VecMath;
+import org.terasology.math.geom.Quat4f;
+import org.terasology.physics.shapes.CollisionShape;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class BulletCompoundShape extends BulletCollisionShape implements org.terasology.physics.shapes.CompoundShape {
+    private final CompoundShape compoundShape;
+    private List<BulletCompoundShapeChild> childList;
+
+    public BulletCompoundShape() {
+        this(new CompoundShape());
+    }
+
+    private BulletCompoundShape(CompoundShape compoundShape) {
+        this.compoundShape = compoundShape;
+        underlyingShape = compoundShape;
+
+        childList = new ArrayList<>();
+    }
+
+    @Override
+    public void addChildShape(org.terasology.math.Transform transform, CollisionShape collisionShape) {
```

Comment:
Can't we provide a com.bulletphysics.linearmath.Transform in input too? Avoiding the toBulletTransform line?

Reply:
Unfortunately, we cannot. The whole point of this refactor is to remove usages of physics engine-specific classes in the engine, and using the Bullet `Transform` class in an engine facing API would defeat that purpose. The whole engine is modified to use `org.terasology.math.Transform`, and it wouldn't make sense to change the accepted type. Furthermore, future inheritors may want to use their own `Transform` representation. It is thus the implementors responsibility to convert the `org.terasology.math.Transform` object into an object of its own liking.